### DISCUSSION TITLE
Use digital ocean space for hosting file upload

### DISCUSSION THREAD

**Discussion by laldevfr** [serverpod/serverpod#593] [Q&A] [answered]
Hi,

To start, congratulations for this V1 and thank you for serverpod which is a very good product.

I want to know if it possible to use [Digital Ocean Space ](https://www.digitalocean.com/products/spaces) (S3 Compatible) for my storage of upload file ? (My serverpod server is already hosted in a digital ocean Droplet)

Laurent.

**Comment by vlidholt** [comment, marked as answer]
If it is S3 compatible, it should be easy to adapt. You may just need to configure the permissions properly. When you get it working it would be awesome if you want to share your experiences!

**Reply by laldevfr** [reply]
After many try it's work. 

Just two thing : 
   - change every endpoints **_$bucket.s3.$region.amazonaws.com_**; by **_$bucket.$region.digitaloceanspaces.com_** into : 
        https://github.com/serverpod/serverpod/tree/main/integrations/serverpod_cloud_storage_s3/lib/src
   - Private not work by defaut, so i
       -  Force to private in file (  req.fields['acl'] = 'private'; ) in every dart file
   With that's it's work great.

Now i try to generate a presign url to get a file.

**Reply by vlidholt** [reply]
Great work!

**Reply by laldevfr** [reply]
For get file, The default client not working and there is no pre-signed method.
So after reading "https://aws.amazon.com/fr/blogs/opensource/introducing-the-aws-sigv4-signer-for-dart/". I code this to get a pre-signed URL to access private file :
```dart
// import 'package:aws_common/aws_common.dart';
// import 'package:aws_signature_v4/aws_signature_v4.dart';
// path : key file
final credentials = AWSCredentials(accessKeyId, secretAccessKey);
final signer = AWSSigV4Signer(credentialsProvider: AWSCredentialsProvider(credentials));

final scope = AWSCredentialScope(region: region,service: AWSService.s3);
final host = '$bucket.$region.digitaloceanspaces.com';
final serviceConfiguration = S3ServiceConfiguration();
final urlRequest = AWSHttpRequest.get(Uri.https(host, path), headers: {AWSHeaders.host: host});
final  signedUrl = await signer.presign(urlRequest, credentialScope: scope,serviceConfiguration: serviceConfiguration, expiresIn: const Duration(minutes: 10));
```
And it's work great for digitalocean spaces 👍
